File: lib/approve.ts
// Promote (or reject) the overwrites staged under _pending/ by the approval gate.
//
// `approve` is the human checkpoint: a gated sync/dump/enrich never overwrites a
// live article that already holds good data — it stages the new version under
// _pending/ and records it in _pending/_manifest.json. This module promotes those
// staged files into the live dump: each replaced live file is first archived to
// _replaced/ (recoverable), then the pending file is moved into place.
//
// Safety default: an edit flagged risky (body-dropped / body-error — i.e. the new
// version would lose or fail to capture a body the live file has) is HELD BACK and
// only promoted when includeRisky is set. So a reformatted-upstream regression can
// never be approved by accident; the reviewer must opt into it explicitly.

import { type Logger, NULL_LOGGER } from "./logger.ts";
import { exists, readJson } from "./fsutil.ts";
import type { Article } from "./track/hashing.ts";
import type { Changelog } from "./changelog.ts";
import {
  archiveReplaced,
  computeRisk,
  livePath,
  loadPendingManifest,
  nowStamp,
  type PendingEntry,
  pendingPath,
  savePendingManifest,
} from "./staging.ts";

export interface ApproveOpts {
  dump: string;
  /** discard the staged files instead of promoting them. */
  reject?: boolean;
  /** restrict to these type dirs (null/undefined = all). */
  typeKeys?: string[] | null;
  /** exclude these type dirs (applied after typeKeys). */
  excludeTypeKeys?: string[] | null;
  /** restrict to these article ids (null/undefined = all). */
  ids?: string[] | null;
  /** archive the replaced live file to _replaced/ before overwriting (default true). */
  archive?: boolean;
  /** also promote edits flagged risky (body-dropped/body-error). */
  includeRisky?: boolean;
  /** preview only: compute + report, change nothing on disk. */
  dryRun?: boolean;
  changelog?: Changelog;
  nowMs: number;
  logger?: Logger;
}

export interface ApproveItem {
  typeKey: string;
  id: string;
  title?: string;
  risk: string[];
  action: "promoted" | "rejected" | "held-risky" | "missing-pending" | "preview";
  archived?: string | null;
}

export interface ApproveResult {
  items: ApproveItem[];
  promoted: number;
  rejected: number;
  heldRisky: number;
  remaining: number; // pending entries still awaiting approval after this op
}

function matches(e: PendingEntry, opts: ApproveOpts): boolean {
  if (opts.typeKeys && opts.typeKeys.length && !opts.typeKeys.includes(e.typeKey)) return false;
  if (opts.excludeTypeKeys && opts.excludeTypeKeys.includes(e.typeKey)) return false;
  if (opts.ids && opts.ids.length && !opts.ids.includes(e.id)) return false;
  return true;
}

export async function approve(opts: ApproveOpts): Promise<ApproveResult> {
  const log = opts.logger ?? NULL_LOGGER;
  const data = await loadPendingManifest(opts.dump);
  const stamp = nowStamp(opts.nowMs);
  const items: ApproveItem[] = [];
  const kept: PendingEntry[] = [];
  let promoted = 0, rejected = 0, heldRisky = 0;

  for (const e of data.entries) {
    if (!matches(e, opts)) {
      kept.push(e);
      continue;
    }
    const pp = pendingPath(opts.dump, e.typeKey, e.id);
    const lp = livePath(opts.dump, e.typeKey, e.id);

    if (!(await exists(pp))) {
      // Manifest entry whose pending file is gone — drop it (don't keep dangling).
      items.push({
        typeKey: e.typeKey,
        id: e.id,
        title: e.title,
        risk: [],
        action: "missing-pending",
      });
      continue;
    }

    // Recompute risk fresh from the actual files so it reflects reality (e.g. after
    // an enrich pass filled the staged article's body).
    let risk: string[] = [];
    let docType = e.typeKey;
    try {
      const pend = await readJson<Article>(pp);
      docType = pend.documentType ?? e.typeKey;
      const live = (await exists(lp)) ? await readJson<Article>(lp) : null;
      risk = computeRisk(live, pend);
    } catch {
      // unreadable pending file -> treat as no risk info
    }

    if (opts.dryRun) {
      items.push({ typeKey: e.typeKey, id: e.id, title: e.title, risk, action: "preview" });
      kept.push(e);
      continue;
    }

    if (opts.reject) {
      await Deno.remove(pp).catch(() => {});
      rejected++;
      items.push({ typeKey: e.typeKey, id: e.id, title: e.title, risk, action: "rejected" });
      continue;
    }

    if (risk.length && !opts.includeRisky) {
      heldRisky++;
      kept.push(e); // stays pending until explicitly included
      items.push({ typeKey: e.typeKey, id: e.id, title: e.title, risk, action: "held-risky" });
      continue;
    }

    // Promote: archive the replaced live file, then move pending into place.
    let archived: string | null = null;
    if (opts.archive !== false) archived = await archiveReplaced(opts.dump, e.typeKey, e.id, stamp);
    await Deno.mkdir(lp.slice(0, lp.lastIndexOf("/")), { recursive: true });
    await Deno.rename(pp, lp);
    promoted++;
    items.push({
      typeKey: e.typeKey,
      id: e.id,
      title: e.title,
      risk,
      action: "promoted",
      archived,
    });
    opts.changelog?.record({
      op: "edited",
      documentType: docType,
      id: e.id,
      title: e.title,
      changed: e.changed,
      hashOld: e.hashOld,
      hashNew: e.hashNew,
      source: "approve",
      detail: archived
        ? `replaced live file archived${risk.length ? ` (risk: ${risk.join(",")})` : ""}`
        : undefined,
    });
  }

  if (!opts.dryRun) {
    data.entries = kept;
    data.generatedAt = new Date(opts.nowMs).toISOString();
    await savePendingManifest(opts.dump, data);
  }
  log.info(
    `approve: promoted=${promoted} rejected=${rejected} held-risky=${heldRisky} ` +
      `remaining=${kept.length}`,
  );
  return { items, promoted, rejected, heldRisky, remaining: kept.length };
}

File: f5kb/lib/approve.py
"""Promote (or reject) overwrites staged under _pending/ by the approval gate."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from f5kb.lib.logger import Logger, NULL_LOGGER
from f5kb.lib.fsutil import path_exists, read_json
from f5kb.lib.staging import (
    PendingEntry,
    archive_replaced,
    change_kind,
    compute_risk,
    diff_parts,
    live_path,
    load_pending_manifest,
    now_stamp,
    pending_path,
    save_pending_manifest,
)


@dataclass
class ApproveItem:
    type_key: str
    id: str
    title: str | None
    risk: list[str]
    changed: list[str]
    action: str  # "promoted"|"rejected"|"held-risky"|"missing-pending"|"preview"
    archived: str | None = None


@dataclass
class ApproveResult:
    items: list[ApproveItem]
    promoted: int
    rejected: int
    held_risky: int
    remaining: int


def _matches(entry: dict, opts: dict) -> bool:
    type_keys = opts.get("type_keys")
    exclude = opts.get("exclude_type_keys") or []
    ids = opts.get("ids")
    if type_keys and entry.get("typeKey") not in type_keys:
        return False
    if entry.get("typeKey") in exclude:
        return False
    if ids and entry.get("id") not in ids:
        return False
    return True


def approve(
    dump: str,
    *,
    reject: bool = False,
    type_keys: list[str] | None = None,
    exclude_type_keys: list[str] | None = None,
    ids: list[str] | None = None,
    archive: bool = True,
    include_risky: bool = False,
    dry_run: bool = False,
    changelog=None,
    now_ms: int,
    logger: Logger = NULL_LOGGER,
) -> ApproveResult:
    log = logger
    data = load_pending_manifest(dump)
    stamp = now_stamp(now_ms)
    items: list[ApproveItem] = []
    kept: list[dict] = []
    promoted = rejected = held_risky = 0
    opts = {
        "type_keys": type_keys,
        "exclude_type_keys": exclude_type_keys,
        "ids": ids,
    }

    for e in data.get("entries") or []:
        if not _matches(e, opts):
            kept.append(e)
            continue

        type_key = e.get("typeKey") or ""
        art_id = e.get("id") or ""
        title = e.get("title")
        pp = pending_path(dump, type_key, art_id)
        lp = live_path(dump, type_key, art_id)

        if not path_exists(pp):
            items.append(ApproveItem(type_key, art_id, title, [], [], "missing-pending"))
            continue

        risk: list[str] = []
        changed: list[str] = e.get("changed") or []
        doc_type = type_key
        try:
            pend = read_json(pp)
            doc_type = pend.get("documentType") or type_key
            live = read_json(lp) if path_exists(lp) else None
            risk = compute_risk(live, pend)
            parts = diff_parts(live, pend)
            if parts:
                changed = parts
        except Exception:
            pass

        if dry_run:
            items.append(ApproveItem(type_key, art_id, title, risk, changed, "preview"))
            kept.append(e)
            continue

        if reject:
            try:
                Path(pp).unlink(missing_ok=True)
            except Exception:
                pass
            rejected += 1
            items.append(ApproveItem(type_key, art_id, title, risk, changed, "rejected"))
            continue

        if risk and not include_risky:
            held_risky += 1
            kept.append(e)
            items.append(ApproveItem(type_key, art_id, title, risk, changed, "held-risky"))
            continue

        # Promote: archive live, move pending into place
        archived: str | None = None
        if archive:
            archived = archive_replaced(dump, type_key, art_id, stamp)
        Path(lp).parent.mkdir(parents=True, exist_ok=True)
        Path(pp).rename(lp)
        promoted += 1
        items.append(ApproveItem(type_key, art_id, title, risk, changed, "promoted", archived))

        notes = [change_kind(changed)]
        if archived:
            notes.append("replaced file archived")
        if risk:
            notes.append(f"risk: {','.join(risk)}")
        if changelog:
            changelog.record(
                "edited", doc_type, art_id,
                title=title, changed=changed,
                hashOld=e.get("hashOld"), hashNew=e.get("hashNew"),
                source="approve", detail="; ".join(notes),
            )

    if not dry_run:
        import time as _t
        import datetime
        data["entries"] = kept
        data["generatedAt"] = datetime.datetime.fromtimestamp(
            now_ms / 1000, tz=datetime.timezone.utc
        ).isoformat().replace("+00:00", "Z")
        save_pending_manifest(dump, data)

    log.info(
        f"approve: promoted={promoted} rejected={rejected} "
        f"held-risky={held_risky} remaining={len(kept)}"
    )
    return ApproveResult(items, promoted, rejected, held_risky, len(kept))
